lib/chef/provider/package/zypper.rb

```diff
@@ -70,7 +71,20 @@ def get_current_versions
         end
 
         def candidate_version
-          @candidate_version ||= package_name_array.each_with_index.map { |pkg, i| available_version(i) }
+          if source_files_exist?
+            logger.trace("#{new_resource} checking rpm status")
+            shell_out!("rpm", "-qp", "--queryformat", "%{NAME} %{VERSION}-%{RELEASE}\n", new_resource.source).stdout.each_line do |line|
+              case line
+              when /^(\S+)\s(\S+)$/
+                current_resource.package_name($1)
+                new_resource.version($2)
+                @candidate_version = $2
+              end
+            end
```

Comment:
the shell_out and its parsing here should probably be extracted out to its own method, see `resolve_source_to_version_obj` in the dnf or yum providers (although you don't have a version obj in this case so probably `resolve_source_to_version`.

Reply:
(in general the dnf provider is a reasonably good template for how the code change should look)